### PR TITLE
trie: fix range prover

### PR DESCRIPTION
**Pull request by rjl493456442**
Fixes the issue described https://github.com/ethereum/go-ethereum/pull/22179#issuecomment-763672300

It's a special case that the trie only has a single trie node and the range proof only contains a single element.

```
TRACE[01-20|15:53:31.702] Fetching range of accounts               id=nice-a reqid=4037200794235010051 root="c2661f…b0463a" origin="000000…000000" limit="0fffff…ffffff" bytes=512.00KiB
TRACE[01-20|15:53:31.702] Delivering range of accounts             id=nice-a reqid=4037200794235010051 hashes=1 accounts=1 proofs=1 bytes=210.00B
panic: interface conversion: trie.node is nil, not *trie.fullNode

goroutine 20 [running]:
github.com/ethereum/go-ethereum/trie.unsetInternal(0xa5c2a0, 0xc00007ed70, 0xc00002b9a0, 0x41, 0x41, 0xc00002b9f0, 0x41, 0x41, 0x20, 0xa5c1a0)
	/home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:302 +0xd65
github.com/ethereum/go-ethereum/trie.VerifyRangeProof(0x6657bd231f66c2, 0x723c83101df021d9, 0xea1a61864a975b5e, 0x3a46b00c81ac76fa, 0xc00014a630, 0x20, 0x20, 0xc000352400, 0x20, 0x20, ...)
	/home/user/go/src/github.com/ethereum/go-ethereum/trie/proof.go:563 +0x7e5
github.com/ethereum/go-ethereum/eth/protocols/snap.(*Syncer).OnAccounts(0xc000147b00, 0xa62960, 0xc00016a3f0, 0x380704bb7b4d7c03, 0xc0003522a0, 0x1, 0x1, 0xc00034c140, 0x1, 0x1, ...)
	/home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync.go:2033 +0x876
github.com/ethereum/go-ethereum/eth/protocols/snap.defaultAccountRequestHandler(0xc00016a3f0, 0x380704bb7b4d7c03, 0x6657bd231f66c2, 0x723c83101df021d9, 0xea1a61864a975b5e, 0x3a46b00c81ac76fa, 0x0, 0x0, 0x0, 0x0, ...)
	/home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:220 +0x11b
created by github.com/ethereum/go-ethereum/eth/protocols/snap.(*testPeer).RequestAccountRange
	/home/user/go/src/github.com/ethereum/go-ethereum/eth/protocols/snap/sync_test.go:162 +0x365
```

### CODE REVIEW

File: trie/proof.go [hunk 1]
```diff
@@ -284,21 +284,34 @@ findFork:
 			return errors.New("empty range")
 		}
 		if shortForkLeft != 0 && shortForkRight != 0 {
+			// The fork point is root node, unset the entire trie
+			if parent == nil {
+				n = nil
```
Comment (holiman) [inline trie/proof.go:289]: Isn't setting this local variable to `nil` a no-op?

File: trie/proof.go [hunk 2]
```diff
@@ -284,21 +284,34 @@ findFork:
...
 			parent.(*fullNode).Children[left[pos-1]] = nil
 			return nil
 		}
 		// Only one proof points to non-existent key.
 		if shortForkRight != 0 {
-			// Unset left proof's path
 			if _, ok := rn.Val.(valueNode); ok {
+				// The fork point is root node, unset the entire trie
+				if parent == nil {
+					n = nil
```
Comment (holiman) [inline trie/proof.go:300]: same

File: trie/proof.go [hunk 3]
```diff
@@ -284,21 +284,34 @@ findFork:
...
 				return nil
 			}
 			return unset(rn, rn.Val, left[pos:], len(rn.Key), false)
 		}
 		if shortForkLeft != 0 {
-			// Unset right proof's path.
 			if _, ok := rn.Val.(valueNode); ok {
+				// The fork point is root node, unset the entire trie
+				if parent == nil {
+					n = nil
```
Comment (holiman) [inline trie/proof.go:312]: same

**Review by holiman — APPROVED** [review]
LGTM, minor comments

**Comment by karalabe** [comment]
Erm, doesn't this PR **always** return `false`?

**Comment by karalabe** [comment]
Oh, there was a true, just in a new return line so it didn't get extra-highlighted. Sry for the noise.

**Comment by rjl493456442** [comment]
@karalabe You can run this test and it will return true https://github.com/ethereum/go-ethereum/pull/22210/files#diff-37d2520502076e40fa878b9a2cf4606e782daefe7c56acbd4eb0b84b0d39c74fR389

**Pull request merged**